sdlib/d/gc/tcache.d

```diff
@@ -54,6 +64,77 @@ public:
 		pd.arena.free(emap, pd, ptr);
 	}
 
+	/**
+	 * Appendables API.
+	 */
+
+	// Get the capacity of the array slice denoted by slice and length.
+	// It includes the space occupied by the slice itself.
+	size_t getArrayCapacity(void* sliceAddr, size_t sliceLen) {
+		if ((sliceAddr is null) || (!sliceLen))
+			return 0;
+
+		auto pd = maybeGetPageDescriptor(sliceAddr);
+		// Return zero if slice is unknown to GC or is to a non-appendable block:
+		if ((pd.extent is null) || (!pd.extent.allocSize))
+			return 0;
+
+		// If an element exists after the slice, capacity is defined as zero.
+		// See also: https://dlang.org/spec/arrays.html#capacity-reserve
+		if (sliceAddr - pd.extent.address + sliceLen < pd.extent.allocSize)
+			return 0;
+
+		// Otherwise, capacity is the block size minus the slice length:
+		return pd.extent.size - sliceLen;
+	}
+
+	// Append array slice denoted by rAddr and rBytes to the one
+	// denoted by lAddr and lBytes, using the former's appendable capacity
+	// if possible, otherwise allocating a new array.
+	void* appendArray(void* lAddr, size_t lBytes, void* rAddr, size_t rBytes) {
```

Comment:
This is the wrong abstraction level, once again. the GC could not care less about arrays. The runtime does. What you want is to be able to extend an existing allocation.

There is no reason for the GC itself to do memcopies and whatnot. All of that is language level semantic that have nothing to do in there.